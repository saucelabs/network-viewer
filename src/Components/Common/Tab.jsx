import PropTypes from 'prop-types';

const Tab = ({ children }) => children;

Tab.propTypes = {
  children: PropTypes.node.isRequired,
};

export default Tab;
