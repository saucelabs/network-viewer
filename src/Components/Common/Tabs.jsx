import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import Styles from './Tabs.styles.scss';

const context = classNames.bind(Styles);

const Tabs = ({
  onUpdate,
  selectedKey,
  defaultSelectedKey,
  children,
  navTabsClassName,
  tabsContainerClassName,
  navLinkClassName,
  activeClassName,
}) => {
  const [items, updateItems] = useState([]);

  useEffect(() => {
    const itemsCollection = [];
    React.Children.forEach(children, (element) => {
      if (React.isValidElement(element)) {
        const {
          name,
          eventKey: key,
          children: component,
        } = element.props;
        itemsCollection.push({
          name,
          key,
          component,
        });
      }
    });
    updateItems(itemsCollection);
  }, [children]);

  const [activeTab, updateTab] = useState(
    defaultSelectedKey || (items && items.length ? items[0].key : null),
  );
  const handleUpdate = (key) => {
    updateTab(key);
    if (onUpdate) {
      onUpdate(key);
    }
  };

  const renderItem = () => {
    const selectedTab = items.find(({ key }) => key === (selectedKey || activeTab));
    return selectedTab ? selectedTab.component : null;
  };

  return (
    <>
      <nav className={context('nav-tabs', navTabsClassName)}>
        {items.map(({ key: item, name }, index) => (
          <a
            key={item}
            className={context('tab-item', navLinkClassName, {
              [activeClassName || Styles.active]: activeTab === item,
            })}
            onClick={() => handleUpdate(item)}
            role="tab"
            tabIndex={index}
          >
            {name}
          </a>
        ))}
      </nav>
      <section className={context(tabsContainerClassName)}>
        {renderItem()}
      </section>
    </>
  );
};

Tabs.propTypes = {
  activeClassName: PropTypes.string,
  children: PropTypes.node.isRequired,
  defaultSelectedKey: PropTypes.string,
  navLinkClassName: PropTypes.string,
  navTabsClassName: PropTypes.string,
  onUpdate: PropTypes.func,
  selectedKey: PropTypes.string,
  tabsContainerClassName: PropTypes.string,
};

Tabs.defaultProps = {
  activeClassName: null,
  defaultSelectedKey: null,
  navLinkClassName: null,
  navTabsClassName: null,
  onUpdate: null,
  selectedKey: null,
  tabsContainerClassName: null,
};

export default Tabs;
