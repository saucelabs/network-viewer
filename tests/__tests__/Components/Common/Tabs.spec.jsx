import React from 'react';
import { shallow } from 'enzyme';

import Tabs from './../../../../src/Components/Common/Tabs';
import Tab from './../../../../src/Components/Common/Tab';

describe('Tabs', () => {
  it('renders without crashing', () => {
    shallow(
      <Tabs>
        <Tab
          key="foo"
          name="Foo"
        >
          <p>Foo Bar</p>
        </Tab>
      </Tabs>,
    );
  });
});
