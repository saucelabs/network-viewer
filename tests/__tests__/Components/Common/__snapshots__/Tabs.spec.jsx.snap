// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders without crashing 1`] = `
<Fragment>
  <nav
    className="nav-tabs"
  />
  <section
    className=""
  />
</Fragment>
`;
